drivers/disk/flashdisk.c

```diff
@@ -47,54 +47,89 @@ static int disk_flash_access_status(struct disk_info *disk)
 	return DISK_STATUS_OK;
 }
 
-static int disk_flash_access_init(struct disk_info *disk)
+static int flashdisk_init_runtime(struct flashdisk_data *ctx,
+				  const struct flash_area *fap)
 {
-	struct flashdisk_data *ctx;
-	const struct flash_area *fap;
 	int rc;
 	struct flash_pages_info page;
+	off_t offset;
 
-	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
-
-	rc = flash_area_open(ctx->area_id, &fap);
-	if (rc < 0) {
-		LOG_ERR("Flash area %u open error %d", ctx->area_id, rc);
-		return rc;
-	}
-
-	k_mutex_lock(&ctx->lock, K_FOREVER);
-
-	disk->dev = flash_area_get_device(fap);
-
-	rc = flash_get_page_info_by_offs(disk->dev, ctx->offset, &page);
+	rc = flash_get_page_info_by_offs(ctx->info.dev, ctx->offset, &page);
 	if (rc < 0) {
 		LOG_ERR("Error %d while getting page info", rc);
-		goto end;
+		return rc;
 	}
 
 	ctx->page_size = page.size;
-	LOG_INF("Initialize device %s", disk->name);
+	LOG_INF("Initialize device %s", ctx->info.name);
 	LOG_INF("offset %lx, sector size %zu, page size %zu, volume size %zu",
 		(long)ctx->offset, ctx->sector_size, ctx->page_size, ctx->size);
 
+	if (ctx->cache_size == 0) {
+		/* Read-only flashdisk, no flash partition constraints */
+		LOG_INF("%s is read-only", ctx->info.name);
+		return 0;
+	}
+
+	if (IS_ENABLED(CONFIG_FLASHDISK_VERIFY_PARTITION_CONSTAINTS)) {
```

Comment:
One more request, please rename this something like `CONFIG_FLASHDISK_VERIFY_PAGE_LAYOUT`, because these may not be constant in case when they are obtained at run-time from flash information, or the SFD.

Reply:
Would it not be better to change all these tests into ASSERTS, then they would be removed anyhow.

Reply:
Please do not use homeopathic ASSERTS. Please consider if it would make sense to move this code block to a `flashdisk_verify_page_layout(struct flashdisk_data *const ctx, ...)` function.

Reply:
ASSERTS are not homeopathic. If the intention of this Kconfig option is to be enabled only during tests this is equivalent of using ASSERTS in a debug mode.

Reply:
ASSERTS are disable per default. We even know about vulnerabilities in the past (e.g. in the diskaccess code) because of wrong assumptions about ASSERTS. This Kconfig option is enabled per default and can be disabled to reduce code size.

Reply:
I would live it as is. We have places where we allow users to enable NULL checks, because once user is sure that everything is OK, the NULL checks may be removed saving some space and time.
If users want it, let them enable it if they do not, let them disable it.

Reply:
I am ok with leaving it as is, but I don't think it is a good idea to introduce Kconfig options to remove some of the code (the same for NULL checks). Some subsystem might be compliant with code rules if some config is set, while it is not compliant int the other case. Just live with the few extra lines of code and try to remove them as a whole.

Reply:
I think asserts are much more suitable here than Kconfig. Basically what this code really does it what the asserts are for.

Reply:
Will failure here, in non-assert version of code, halt the device completely? Or will just issue errors/warnings and allow the rest of the system work?

Reply:
No, in non-assert version it will just "work" in the same way as it does with the Kconfig disabled.

Reply:
And with Kconfig enabled?

Reply:
Well, if it gets converted to asserts, there's no need for extra Kconfig option so it would be removed.

The assert enabled would be equivalent to Kconfig enabled with the difference that it won't let system execute (i.e. fatal failure) if the constraints are not met. The fatal error here is really correct, as the problem is with the *code* (application, configuration, or whatever) itself.

The assert disabled would be equivalent to Kconfig disabled, i.e. no warning if the constraints are not met.

Reply:
Yes, but in case when you have failed to test this part or accidentally have uploaded wrong image, your device is dead and halt in secondary subsystem prevents you from retreating from mistake by uploading good firmware.

On the other side if there are no asserts enabled, but you have uploaded image to wrong device or there is some other failure in flash disk subsystem the device will still work, with errors, but will work and will allow you to fix your mistake.

Reply:
I don't agree. If the constraints are not met, it is software developer fault and you cannot really draw any conclusions what the **essentially flawed** software will do. There is no cure for sloppy development. Asserts are really designed to be used for such cases, i.e. checking something that **must** be true and can only be not true during development if the developer doesn't realize that there are some constraints.

Reply:
??? Any failure to get a system working should be possible to correct by uploading a new image or restoring the previous image. If the bootloader does not provide this there is something seriously wrong in your setup. A device that works with errors can create a locked up situation that you cannot recover from so it is bad to try and support this scenario.

Reply:
I think either this should be either a flashdisk feature that is always there (i.e. cannot be disabled in any way, only the log messages can be removed from binary) or this is changed into asserts. In other words, the use of custom Kconfig for disabling this sounds inappropriate.

Reply:
We are talking about difference between:
```
Init.. something is not right with flash disk... log errors... notify remote control... not confirm image... reboot... revert
```
and
```
Init... BLERP! IAMDEAD! HALT!... (maybe reboot and revert)
```